src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -393,13 +379,7 @@ private DocumentState SetParseOptions(ParseOptions options, bool onlyPreprocesso
 
                 SyntaxTree? newTree = null;
 
-                if (existingTree is IRecoverableSyntaxTree recoverableTree &&
-                    !recoverableTree.ContainsDirectives)
-                {
-                    // It's a recoverable tree, so we can try to reuse without even having to need the root
-                    newTree = recoverableTree.WithOptions(options);
-                }
-                else if (existingTree.TryGetRoot(out var existingRoot) && !existingRoot.ContainsDirectives)
+                if (existingTree.TryGetRoot(out var existingRoot) && !existingRoot.ContainsDirectives)
```

Comment:
@jasonmalinowski this it eh most interesting part.  We no longer have a place where we're storing this 'ContainsDirective' bit as we're not using recoverable trees.

I recommend the following: we have a cache somewhere with a `CWT<SyntaxTree, AsyncLazy<bool>>` that computes and caches if a tree contains directives.  we can then consult that as necessary to make thsi determination.  When we fork a tree and do not change teh root, we should put that tree in here up front.  similarly, if we change the root to a new known root, we can synchronously update this by checking the .ContainsDirectives bit of the new root.

Reply:
Chatting with @CyrusNajmabadi we suspect there's a follow-up we can do to delete more stuff here.